src/mbgl/gl/attribute.cpp

```diff
@@ -1,47 +1,19 @@
 #include <mbgl/gl/attribute.hpp>
 #include <mbgl/gl/context.hpp>
-#include <mbgl/gl/defines.hpp>
 
 namespace mbgl {
 namespace gl {
 
 using namespace platform;
 
-void bindAttributeLocation(Context& context, ProgramID id, AttributeLocation location, const char* name) {
```

Comment:
This function's task is to set an explicit binding for an attribute location before (re-)linking the program. The code flow is a bit hard to follow due to it's heavily templated nature, but it basically worked like this:

- Link program
- Query program to find out what attributes are actually present. This is necessary because we disable/enable attributes for data-driven styling (DDS)
- The `AttributeLocations` "iterates" (using template expansions...) over an explicitly ordered list of attributes (they're stored in the `*_program.hpp` in a `TypeList`: https://github.com/mapbox/mapbox-gl-native/blob/8a3d07fcd83d01415b672dd5f45a5e08af7d5594/src/mbgl/programs/circle_program.hpp#L18-L19
    and augmented in the `gfx::Program` class with the attributes for data-driven styling, which is a subset of this list: https://github.com/mapbox/mapbox-gl-native/blob/8a3d07fcd83d01415b672dd5f45a5e08af7d5594/src/mbgl/style/layers/circle_layer_properties.hpp#L61-L73
  That means that attributes get sequentially numbered, starting with `location = 0` for the second linking pass. Some (many?) OpenGL implementations have massive performance issues if the attribute 0 is disabled (e.g. because it is not used). With this weird double-linking approach, we're ensuring that location 0 is always enabled because it is the first attribute of the program's non-DDS attribute list. Removing relinking is definitely possible, but we should preserve the current behavior of explicitly assigning attribute bindings with `glBindAttribLocation`. We could probably get away with only running it for the first attribute in the list, since we don't really care whether the rest are numbered sequentially. What doesn't work is to run `glBindAttribLocation` blindly for all attributes because we sometimes have more than 8 of those, and many implementations don't support attribute locations beyond 7, so we have to have some awareness of which attributes are in use or not. We could query this information from the `gfx::AttributeBindings<AttributeList>` object passed to the `draw` function (that we also use to compute the `gl::AttributeKey<AttributeList>` for finding the correct program instance in the first place). Explicitly binding all attributes before the first link also has the advantage that we don't have to query the attribute locations after linking because they're guaranteed to be the ones we set them to.

Reply:
> Query program to find out what attributes are actually present. This is necessary because we disable/enable attributes for data-driven styling (DDS)

> we don't have to query the attribute locations after linking because they're guaranteed to be the ones we set them to.

The patch is replacing usage of glGetProgramiv and per-attrib glGetActiveAttrib by glGetAttribLocation. I don't think it is introducing redundancy.


> With this weird double-linking approach, we're ensuring that location 0 is always enabled because it is the first attribute of the program's non-DDS attribute list. Removing relinking is definitely possible, but we should preserve the current behavior of explicitly assigning attribute bindings with glBindAttribLocation. We could probably get away with only running it for the first attribute in the list, since we don't really care whether the rest are numbered sequentially. 

This proposal sounds good and looks safe to assert (I checked the shaders code before working on this) position attribute being always enabled, for all the shaders on index 0. I'll update patch as that would avoid double linking and resolve other concerns.

Reply:
> I don't think it is introducing redundancy.
I didn't mean to suggest that your patch does; I was describing the old approach.

Reply:
The patch [Remove re-linking: bind position attribute at location 0.](https://github.com/mapbox/mapbox-gl-native/pull/14482/commits/99e5f262645aeff19797cc1ed87b023add895e59) binds the position  related attribute at location 0. 

To expand on the comment:
    // It is important to have attribute at position 0 enabled: conveniently,
    // position attribute is always first and always enabled. The integrity of
    // this assumption is verified in AttributeLocations::queryLocations and
    // AttributeLocations::getFirstAttribName.

there is compile time check that the first element is position and assert verifying the bound location 0 enabled. 

I have [measured the time spent just on relinking](https://gist.github.com/astojilj/29bd5a5c5dc0b2d9f29ecb660da07fbf) using release build on iPhone SE (A9, same as iPhone 6S):

- 1st run after reboot or installation
Total 37.14ms, average per program:1.86ms

- reopening
Total: 2.47ms, average per program: 0.12ms

This time we save using the patch here.